Client and Apartment interface logic

AnnYaniv opened the pull request:
Register a new client
Open a new apartment to manage (can be of different capacity)
List clients sorted by ID, name, status
List apartments sorted by ID, price, capacity, availability
Check in a client to an apartment
Check out a client from an apartment
Adjust the price of an apartment
Calculate the current price for the client’s stay

alexeyburak committed:
add domain storage logic

alexeyburak committed:
add domain use cases

alexeyburak committed:
add apartment service logic

alexeyburak committed:
add client service logic

alexeyburak committed:
add base command and exception handlers

AnnYaniv committed:
Refactor
ErrorMessages from class to enum
Services and repositories to singleton
Move comparator from map to enum
Add another method for Commands(experimental)

AnnYaniv committed:
Configure app
Change pom.xml
Add logback configuration

alexeyburak committed:
Add In Memory Storage Unit tests (#1)

Add storage tests

---------

Co-authored-by: AnnYaniv <yaniv.anya@gmail.com>

alexeyburak committed:
Add Application Commands Logic (#2)

* Add commands interface
* Separate EntityFactory to ClientFactory and ApartmentFactory
* Add adjust command logic
* Add list of command
* Refactor readme file
* Add CrudRepository interface
* Changed client service and repository logic
* Creating a model package
* Change repos and services logic
* Replacing repos to services in commands

---------

Co-authored-by: AnnYaniv <yaniv.anya@gmail.com>
Co-authored-by: Aibar <aibar.west@gmail.com>
Co-authored-by: Ali Aliyev <alialiyevlog@gmail.com>

AnnYaniv committed:
Fix check-in check-out

alexeyburak committed:
refactor services, repositories, add crud use cases

AnnYaniv committed:
Refactor
Get rid of singleton dependencies
Add crud-interfaces
Separate commands to respective packages

AnnYaniv committed:
Tests
Add new dependency
Add service and repo tests

alexeyburak committed:
clean up code

AnnYaniv committed:
Test ApartmentRepository and ApartmentService
Fix adjust method

AnnYaniv committed:
Refactoring tests
Rename StoreTest in RepositoryTest
Change assertThrows to assertThatThrownBy

AnnYaniv committed:
Change assertTrue to assertThat(...).isTrue()

AnnYaniv committed:
Merge pull request #4 from alexeyburak/apartment-unit-tests

Unit tests for Apartment-classes

alexeyburak committed:
Add GitHub Actions workflow (#3)

alexeyburak committed:
Add unit tests for client repository and service (#5)

* add client repo and service unit tests

alexeyburak committed:
add dependency to run tests with maven

maximtereshchenko commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,46 @@
+
+jobs:
+
+  build:
```
Я бы на вашем месте разделил эти две задачи: сборка на пуш, тесты на ПР. Но в целом - огонь!

maximtereshchenko commented on README.md:
```diff
@@ -1,3 +1,135 @@
\ No newline at end of file
+This is a project for Java intensive course.
+
+# Application Commands
```
Хотелось бы команду help, которая я бы печатала список доступных команд

maximtereshchenko commented on pom.xml:
```diff
@@ -12,16 +12,102 @@
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <version>3.24.2</version>
```
Обычно в проектах принято все версии прятать в properties. Хотя я лично не вижу большой выгоды

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/application/CustomErrorMessage.java:
```diff
@@ -0,0 +1,24 @@
+
+@AllArgsConstructor
+@Getter
+public enum CustomErrorMessage {
```
Не совсем вижу смысл в этом классе, если эти сообщения используются только в одном месте. Можно было бы их там написать напрямую.

maximtereshchenko replied:
Хотя с другой стороны удобно иметь все в одном месте

maximtereshchenko replied:
И еще вы ничего не выигрываете от enum здесь. Можно было бы обойтись просто 
`public static final String APARTMENT_IS_ALREADY_EXISTS = "Apartment is already exists";`

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/application/command/ApplicationCommand.java:
```diff
@@ -0,0 +1,23 @@
+
+@AllArgsConstructor
+@Getter
+public enum ApplicationCommand {
```
Тоже не уверен, что здесь необходим enum

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/application/command/ArgumentsValidator.java:
```diff
@@ -0,0 +1,8 @@
+import java.util.List;
+
+@FunctionalInterface
+public interface ArgumentsValidator<T> {
```
Не совсем понятно, зачем нужен интерфейс, если классы его реализующие пользуются методом для своих нужд. Было бы логично сделать абстрактный класс, который проводил проверку переданных аргументов перед выполнением команды

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/application/command/CommandStarter.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    private void execute(Command command, List<String> args) {
+        new BusinessExceptionHandlingCommand(
```
Было бы логичнее декорировать команды в месте создания CommandStarter. Использование здесь их добавляет дополнительную причину для изменения этому классу

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/application/command/GenericExceptionHandlingCommand.java:
```diff
@@ -0,0 +1,31 @@
+    public void execute(PrintStream output, List<String> arguments) {
+        try {
+            original.execute(output, arguments);
+        } catch (IllegalArgumentException | IndexOutOfBoundsException e) {
```
Мне казалось, что вы проверяете длину аргументов с помощью ArgumentsValidator. Откуда может вылезти IndexOutOfBoundsException?

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/application/command/additional/CalculateClientStayCurrentPriceCommand.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Override
+    public void execute(PrintStream output, List<String> arguments) {
+        double price = useCase.calculatePrice(UUID.fromString(arguments.get(1)));
```
НИКОГДА НЕ ИСПОЛЬЗУЙТЕ DOUBLE ДЛЯ ДЕНЕГ!!!!!!

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/application/command/additional/ExitApplicationCommand.java:
```diff
@@ -0,0 +1,23 @@
+
+public final class ExitApplicationCommand implements Command {
+
+    private static final String EXIT_MESSAGE = "Goodbye, have a nice day!";
```
Зачем нужна константа, если она используется только в одном месте?

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/application/command/crud/CreateEntityCommand.java:
```diff
@@ -0,0 +1,43 @@
+        switch (chosenEntity) {
+            case APARTMENT -> apartmentService.save(ApartmentFactory.createApartment(trimmed));
+            case CLIENT -> clientService.save(ClientFactory.createClient(trimmed));
+            default -> throw new IllegalArgumentException(CustomErrorMessage.UNKNOWN_ENTITY.getMessage());
```
С Java 17 вы получите ошибку компиляции, если в switch будет не хватать веток для enum. А с default - будет ошибка во время выполнения

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/model/Apartment.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.UUID;
+
+@Data
+public final class Apartment {
```
а нельзя record использовать?

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/service/factory/ApartmentFactory.java:
```diff
@@ -0,0 +1,66 @@
+    private ApartmentFactory() {}
+
+    public static Apartment createApartment(List<String> args) {
+        return args.isEmpty() ? createRandomApartment() : createApartmentFromArgs(args);
```
Выглядит не очень, в бизнес логике не было требования создавать что-то случайное

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/service/impl/ApartmentService.java:
```diff
@@ -0,0 +1,91 @@
+
+public final class ApartmentService implements AdjustApartmentPriceUseCase,
+        ListApartmentsUseCase, CrudService<Apartment, ApartmentEntity> {
+    private static final Logger LOG = LoggerFactory.getLogger("service-logger");
```
обычно название класса используют в качестве названия логгера

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/service/impl/ApartmentService.java:
```diff
@@ -0,0 +1,91 @@
+    private static final Logger LOG = LoggerFactory.getLogger("service-logger");
+
+    private final SortableCrudRepository<Apartment, ApartmentSort> store;
+    private final Function<Apartment, ApartmentEntity> toEntityMapper = (apartment ->
```
а нельзя было метод просто сделать.

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/service/impl/ApartmentService.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    @Override
+    public boolean has(UUID id) {
```
Зачем этому методу быть частью интерфейса если он используется только в самом классе?

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/service/impl/ClientService.java:
```diff
@@ -0,0 +1,129 @@
+
+    @Override
+    public void checkIn(UUID clientId, UUID apartmentId) {
+        ClientEntity client = getById(clientId);
```
а что если клиент BANNED?

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/service/impl/ClientService.java:
```diff
@@ -0,0 +1,129 @@
+    @Override
+    public void checkIn(UUID clientId, UUID apartmentId) {
+        ClientEntity client = getById(clientId);
+        ApartmentEntity apartment = apartmentService.getById(apartmentId);
```
а что если апартаменты уже заняты?

maximtereshchenko commented on src/main/java/com/andersenlab/hotel/service/impl/ClientService.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    @Override
+    public void save(Client client) {
```
Интерфейс кривой. Я передаю в метод готового клиента со статусом и апартаментами, а метод это все игнорирует. Interface segregation principle violation

maximtereshchenko commented on src/test/java/com/andersenlab/hotel/repository/inmemory/InMemoryApartmentRepositoryTest.java:
```diff
@@ -0,0 +1,157 @@
+class InMemoryApartmentRepositoryTest {
+
+    private InMemoryApartmentRepository target;
+    private Apartment apartment1, apartment2, apartment3, apartment4, apartment5;
```
Брррр, нынче так никто не делает

maximtereshchenko commented on src/test/java/com/andersenlab/hotel/repository/inmemory/InMemoryApartmentRepositoryTest.java:
```diff
@@ -0,0 +1,157 @@
+        target.save(apartment1);
+        Optional<Apartment> actual = target.getById(apartment1.getId());
+
+        assertThat(actual).isPresent().contains(apartment1);
```
зачем делать isPresent, если делаете contains?

maximtereshchenko commented on src/test/java/com/andersenlab/hotel/repository/inmemory/InMemoryApartmentRepositoryTest.java:
```diff
@@ -0,0 +1,157 @@
+        apartment1 = new Apartment(UUID.randomUUID(), BigDecimal.ONE, BigInteger.ONE,
+                true, ApartmentStatus.AVAILABLE);
+
+        apartment2 = new Apartment(UUID.randomUUID(), BigDecimal.TWO, BigInteger.TWO,
```
у меня BigDecimal.TWO не компилируется

maximtereshchenko commented on src/test/java/com/andersenlab/hotel/repository/inmemory/InMemoryClientRepositoryTest.java:
```diff
@@ -0,0 +1,149 @@
+
+        client1 = new Client(UUID.randomUUID(), "name-1", ClientStatus.NEW);
+
+        client2 = new Client(UUID.randomUUID(), "name-2", ClientStatus.BANNED);
```
не рекомендую в тестах использовать случайные значения

maximtereshchenko commented on src/test/java/com/andersenlab/hotel/repository/inmemory/InMemoryClientRepositoryTest.java:
```diff
@@ -0,0 +1,149 @@
+
+        client2 = new Client(UUID.randomUUID(), "name-2", ClientStatus.BANNED);
+
+        client3 = new Client(UUID.fromString("9d76f43e-4ce2-40ba-90e1-37a6bc425ee1"),
```
Не очень читабельно. Лучше было бы 0000-000-00-0001 (как-то так)

maximtereshchenko commented on src/test/java/com/andersenlab/hotel/service/ApartmentServiceIntegrationTest.java:
```diff
@@ -0,0 +1,4 @@
+package com.andersenlab.hotel.service;
+
+public class ApartmentServiceIntegrationTest {
```
пусто :(

maximtereshchenko commented on src/test/resources/junit-platform.yml:
```diff
@@ -0,0 +1,10 @@
+        classes:
+          default: concurrent
+      config:
+        strategy: dynamic
```
Идея изначально ориентировать на тесты бегущие параллельно - это неплохо. Но в вашем случае, боюсь, оно даже медленнее будет выполняться, из-за доп расходов на многопоточность

maximtereshchenko reviewed (approved):
Ребята, мне в целом понравилось. Конечно, не идеально. Но в дальнейшем, мне кажется, будет просто вносить изменения

AnnYaniv committed:
Test fixes
Deleting random values from tests

AnnYaniv committed:
Fixing double for price values
BigDecimal used instead

AnnYaniv committed:
Fix ApartmentService
Using method reference instead of function

AnnYaniv committed:
Creating help-command

stunt23 committed:
create command decorator
change exitMessage

AnnYaniv committed:
Delete IndexOutOfBoundsException

stunt23 committed:
fix imports

AnnYaniv committed:
Merge branch 'command-refactoring' into apartment-refactor

# Conflicts:
#	src/main/java/com/andersenlab/hotel/Main.java
#	src/main/java/com/andersenlab/hotel/application/command/BusinessExceptionHandlingCommand.java
#	src/main/java/com/andersenlab/hotel/application/command/CommandStarter.java
#	src/main/java/com/andersenlab/hotel/application/command/GenericExceptionHandlingCommand.java

AnnYaniv committed:
Change shown message in CommandStarter
Add help command to CommandCreator
Add private constructor to CommandsCreator
Change message in CommandStarter

alexeyburak committed:
Fix services and commands (#7)

* fix workflows
* fix junit config
* delete randomEntitiesCreators in fabrics
* checkIn logic for BANNED client and Function to method reference
* Interface segregation principle
* fix validator interface

AnnYaniv committed:
refactor ClientService and CommandsCreator

AnnYaniv committed:
Merge pull request #9 from alexeyburak/apartment-refactor

Refactor Apartment-classes and command-classes

alexeyburak committed:
Add check for reserved apartment (#11)
